### PR TITLE
Support post requests in the frontend queryrange handler.

### LINKED ISSUE #2020: Loki Grafana Alerts are not working via Prometheus datasource
**What happened**:
In Loki we don't have yet an alerting storing, so we took advantages of the similarities between Loki and Prometheus API to create Grafana Alert on top of Loki but added as a Prometheus datasource, see below.

image

We even have a video explaining how to do that https://www.youtube.com/watch?v=GdgX46KwKqo

Now it seems that since 6.6 alert are failing to be evaluated, although graphing still works see below:

image

It seems that something is now parsing the PromQL query and since Loki has some differences it fails. Although the query is actually the same as PromQL (`sum(rate({cluster="dev-us-central1", job="loki-boltdb-shipper/distributor", level!~"debug|info"}[5m])) by (level)`).

Response from the test endpoint:

```
{"firing":true,"state":"pending","conditionEvals":" = true","timeMs":"12.647ms","error":"tsdb.HandleRequest() error bad_response: readObjectStart: expect { or n, but found p, error found in grafana/grafana#1 byte of ...|parse error|..., bigger context ...|parse error : syntax error: unexpected $end|...","logs":[{"message":"Condition[0]: Query","data":{"from":1588256991975,"queries":[{"refId":"A","model":{"expr":"sum(rate({cluster=\"dev-us-central1\", job=\"loki-boltdb-shipper/distributor\", level!~\"debug|info\"}[5m])) by (level)","interval":"","intervalFactor":3,"legendFormat":"{{level}}","refId":"A"},"datasource":{"id":3,"name":"Loki-Prometheus"},"maxDataPoints":0,"intervalMs":0}],"to":1588257291975}}]}
```

**What you expected to happen**:

Alert to still works with Loki Prometheus datasource.

**How to reproduce it (as minimally and precisely as possible)**:

See the video.

**Environment**:
- Grafana version: seems to break from 6.6 and forward
- Data source type & version: Loki as Prometheus datasource

/cc @aocenas

### PR DESCRIPTION
**Pull request by cyriltovena**
Signed-off-by: Cyril Tovena <cyril.tovena@gmail.com>

I also took the opportunity to refactor a bit the code so that it's easier to write test and check which roundtripper is called based on a query.

I think ultimately we need to move `r.ParseForm` in the `loghttp.Parse....` functions, because it's in multiple places:
- querier middleware
- frontend codec
- frontend rountripper.

Fixes #2020

### CODE REVIEW

File: pkg/querier/queryrange/roundtrip.go [hunk 1]
```diff
@@ -55,60 +54,87 @@ func NewTripperware(cfg Config, log log.Logger, limits Limits, registerer promet
...
+// transformRegexQuery backport the old regexp params into the v1 query format
+func transformRegexQuery(req *http.Request, expr logql.LogSelectorExpr) logql.LogSelectorExpr {
+	regexp := req.Form.Get("regexp")
+	if regexp != "" {
+		expr = logql.NewFilterExpr(expr, labels.MatchRegexp, regexp)
+		params := req.URL.Query()
+		params.Set("query", expr.String())
+		req.URL.RawQuery = params.Encode()
+		// force the form and query to be parsed again.
+		req.Form = nil
```
Comment (owen-d) [inline pkg/querier/queryrange/roundtrip.go:123]: I'm a bit scared of setting this to nil here. I'm afraid it exposes us to issues extending this in the future -- it's easy to think the form would be parsed already. That being said, I think we should merge this. It's just something we should think about improving.

Comment (cyriltovena) [reply pkg/querier/queryrange/roundtrip.go:123]: Yeah I don't have a solution yet for that. The problem is that ParseForm does nothing if it's already populated and we need to update the querystring.

Comment (cyriltovena) [reply pkg/querier/queryrange/roundtrip.go:123]: If you remove this a test will fail.

**Comment by JnMik** [comment]
Very nice so we will be able to do Loki alerting with this  in 6.7.4 right ? 
 with prometheus-loki datasource 
thanks mate !

**Comment by cyriltovena** [comment]
Yes not sure when grafana change to post but this fixes alerts issue.

**Pull request merged**